pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,423 @@
+		stateRootInHeader: chain.GetConfig().StateRootInHeader,
+		shutdownCallback:  shutdownCallback,
+
+		quit:                  make(chan struct{}),
+		exiterToOIDDownloader: make(chan struct{}),
+		exiterToShutdown:      make(chan struct{}),
+		oidDownloaderToExiter: make(chan struct{}),
+		blockPutterToExiter:   make(chan struct{}),
+
+		// Use buffer of two batch sizes to load OIDs in advance:
+		//  * first full block of OIDs is processing by Downloader
+		//  * second full block of OIDs is available to be fetched by Downloader immediately
+		//  * third half-filled block of OIDs is being collected by OIDsFetcher.
+		oidsCh: make(chan oid.ID, 2*cfg.OIDBatchSize),
+
+		// Use buffer of a single OIDs batch size to provide smooth downloading and
+		// avoid pauses during blockqueue insertion.
+		blocksCh: make(chan *block.Block, cfg.OIDBatchSize),
+	}, nil
+}
+
+// Start runs the block fetcher service.
+func (bfs *Service) Start() error {
+	if !bfs.isActive.CompareAndSwap(false, true) {
+		return nil
+	}
+	bfs.log.Info("starting NeoFS BlockFetcher service")
+
+	var err error
+	bfs.client, err = neofs.GetSDKClient(context.Background(), bfs.cfg.Addresses[0], 10*time.Minute)
+	if err != nil {
+		bfs.isActive.CompareAndSwap(true, false)
+		return fmt.Errorf("create SDK client: %w", err)
+	}
+
+	// Start routine that manages Service shutdown process.
+	go bfs.exiter()
+
+	// Start OIDs downloader routine.
+	go bfs.oidDownloader()
+
+	// Start the set of blocks downloading routines.
+	for i := 0; i < bfs.cfg.DownloaderWorkersCount; i++ {
+		bfs.wg.Add(1)
+		go bfs.blockDownloader()
+	}
+
+	// Start routine that puts blocks into bQueue.
+	go bfs.blockPutter()
+
+	return nil
+}
+
+// oidDownloader runs the appropriate blocks OID fetching method based on the configuration.
+func (bfs *Service) oidDownloader() {
+	defer close(bfs.oidDownloaderToExiter)
+
+	var err error
+	if bfs.cfg.SkipIndexFilesSearch {
+		err = bfs.fetchOIDsBySearch()
+	} else {
+		err = bfs.fetchOIDsFromIndexFiles()
+	}
+
+	if err != nil {
+		bfs.log.Error("NeoFS BlockFetcher service: OID downloading routine failed", zap.Error(err))
+	}
+	// Gracefully stop the service since there's nothing to do anymore.
+	bfs.stopService()
+}
+
+func (bfs *Service) blockPutter() {
+	defer close(bfs.blockPutterToExiter)
+	for b := range bfs.blocksCh {
+		err := bfs.putBlock(b)
+		if err != nil {
+			bfs.log.Error("failed to put block", zap.Error(err))
+			bfs.stopService()
+			return
+		}
+	}
+}
+
+func (bfs *Service) blockDownloader() {
+	defer bfs.wg.Done()
+
+	for blkOid := range bfs.oidsCh {
+		ctx, cancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+		defer cancel()
+
+		rc, err := bfs.objectGet(ctx, blkOid.String())
+		if err != nil {
+			bfs.log.Error("failed to objectGet block", zap.Error(err))
+			bfs.stopService()
+			return
+		}
+
+		b, err := bfs.getBlock(rc)
+		if err != nil {
+			bfs.log.Error("failed to get block", zap.Error(err))
+			bfs.stopService()
```

Comment:
Let's rename `bfs.getBlock` to `bfs.readBlock` since this method reads the block from readcloser. And adjust error correspondingly.